### PR TITLE
fix: add types

### LINKED ISSUE #1: Add types
Hey hello, I could give a try and and create a pr for adding types to this repo if you are interested or if someone else is not already on it

### PR DESCRIPTION
**Pull request by gkampitakis**
closes #1 . Added the types and the tests as discussed. I am not famliar with types testing so might have missed something

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+     * @default 10000
+     */
+    delay: number
+  }
+
+  type Signals = 'SIGTERM' | 'SIGINT';
+
+  type CloseWithGraceCallback = (
+    { err, signal, manual }: { err: Error, signal: Signals, manual: boolean },
+    cb: (error?: Error) => void
```
Comment (mcollina) [inline index.d.ts:14]: the callback is optional. In that case the function should return a promise. If there is a callback, the function must not return a promise.

File: index.test-d.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { expectType, expectAssignable } from 'tsd'
+import * as closeWithGrace from '.'
+import { Options, CloseWithGraceCallback, Signals } from '.'
+
+expectAssignable<CloseWithGraceCallback>((param: { err: Error, manual: boolean, signal: Signals }, cb) => Promise.resolve())
```
Comment (mcollina) [inline index.test-d.ts:5]: this should not be possible.

**Review by mcollina — CHANGES_REQUESTED** [review]

**Comment by mcollina** [comment]
Good work!

File: index.d.ts (hunk 1 above)
Comment (gkampitakis) [reply index.d.ts:14]: You are totally right but can't find a way to achieve that 😞. I tried overloading but seems like creating a union of the  types 

```ts
type asyncAddCb = (result: number) => void
// define all valid function signatures
function asyncAdd (a: number, b: number): Promise<number>
function asyncAdd (a: number, b: number, cb: asyncAddCb): void

function asyncAdd (a: number, b: number, cb?: asyncAddCb) {
  const result = a + b

  return Promise.resolve(result);
}
```

also gave a try to this syntax 
`typeof cb extends undefined ? Promise<void> : void;` but doesn't seem to be correct 😕

Not sure if someone is a ts expert to share some knowledge here 😅

Comment (mcollina) [reply index.d.ts:14]: At least let's not test the bad behavior.

File: index.test-d.ts (hunk 2 above)
Comment (gkampitakis) [reply index.test-d.ts]: I changed the tests at least to match with the correct expectations is this what you had in mind ? 

```ts
expectAssignable<CloseWithGraceCallback>(async (param: { err: Error, manual: boolean, signal: Signals }) => { })
expectAssignable<CloseWithGraceCallback>((param: { signal: Signals }, cb) => { })
```

also I did some more investigation no luck that was the closer I got but couldn't port it to the `.d.ts`
[TS playground](KYDwDg9gTgLgBDAnmYcDCAjOBeOAKYKKaAfgC44BRI6AShwD44A3CASwBMBuAKFElgJkqAAoBDKGIC2AZxxwA3jzgq4hKOSo0ovVXCliAdgFcxAG00YIEM8CO7VMtgHND5zQHIAygEkA4gAqlABKALIecAA)

File: index.d.ts (hunk 1 above)
Comment (fox1t) [reply index.d.ts:14]: @gkampitakis actually you overload works fine as you can see here on the [playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAhgziAdgYwIIBN0GEBGUC8UAFAE4RwCuANsAFxSIUC2OEJAlAQHxQBuA9gEt0AKABmFFMEH9EsBCgzpiMeoxZsANFBxrmrDvQAKJfk0FwIAHnUGu4ycmmz5SNJhV6NJbbob6tKGQ-eDclXHZ6AWERBykZOVDFDyJVf29fLwNtYIB+eiT3bBxOAG8RKCDZOGAoMkoaAlgoAGodCqhBMWJgso7KyuDScmpgdn6BsmAKEkQOgF8OqZm5EzMLCAA6ev4qXghhhrGAbhFF1IVkKAl4lyI+garEGrqRxsIYAHcYQVrCpSIAFZNAA2caPZDVXZbKj8ADmh1G41iKIh1Vq9VGACYmv9MEDQdpETROPgeOVHmjntDNrCEZiSRN5uNmfcREA).

image

image

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+     * The numbers of milliseconds before abruptly close the process
+     * @default 10000
+     */
+    delay: number
+  }
+
+  type Signals = 'SIGTERM' | 'SIGINT';
+
+  type CloseWithGraceCallback = (
+    { err, signal, manual }: { err: Error, signal: Signals, manual: boolean },
```
Comment (fox1t) [inline index.d.ts:13]:
```suggestion
    { err, signal, manual }: { err?: Error, signal?: Signals, manual?: boolean },
```

File: index.test-d.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import { expectType, expectAssignable } from 'tsd'
+import * as closeWithGrace from '.'
+import { Options, CloseWithGraceCallback, Signals } from '.'
+
+expectAssignable<CloseWithGraceCallback>(async (param: { err: Error, manual: boolean, signal: Signals }) => { })
```
Comment (fox1t) [inline index.test-d.ts:5]: We need to test also partial parameters being passed since the run function pass only one parameter per call.

**Review by fox1t — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 1 above)
Comment (gkampitakis) [reply index.d.ts:14]: When I try to do this though 
```
type CloseWithGraceCallback = {
    (
      { err, signal, manual }: { err?: Error, signal?: Signals, manual?: boolean },
      cb: (error?: Error) => void
    ): void
    (
      { err, signal, manual }: { err?: Error, signal?: Signals, manual?: boolean },
    ): Promise<void>
    (
      { err, signal, manual }: { err?: Error, signal?: Signals, manual?: boolean },
      cb?: (error?: Error) => void
    ): Promise<void> | void
  }
```

I get those two errors from the callback case 
Screenshot 2021-03-01 at 09 37 36
Screenshot 2021-03-01 at 09 37 53

**Comment by fox1t** [comment]
@gkampitakis I figured out how to accomplish what you wanted. 
Some points to remember:
1. if a function returns `void` the implementer can return every type and the checks will pass. You can see it as `any` but you can't assign it. It exists for the caller and it literally means "You should don't care about the value my returned value"
2. we need to have 2 different interfaces one for the async version and the other for the callback one. In fact, overloads are intersections and not unions and the passed function should implement both versions at the same time (this is impossible).
I created 2 gists, feel free to use them to "adjust" the typings.

- https://gist.github.com/fox1t/7be1b459c71d87a10c21ffa73584097a
- https://gist.github.com/fox1t/6206469e855c350ccf69c3e6255f4357

:)

**Comment by gkampitakis** [comment]
> @gkampitakis I figured out how to accomplish what we need.
> Some points to remember:
> 
> 1. if a function returns `void` the implementer can return every type and the checks will pass. You can see it as `any` but you can't assign it. It exists for the caller and it literally means "You should don't care about the value my returned value"
> 2. we need to have 2 different interfaces one for the async version and the other for the callback one. In fact, overloads are intersections and not unions and the passed function should implement both versions at the same time (this is impossible).
>    I created 2 gists, feel free to use them to "adjust" the typings.
> 
> * https://gist.github.com/fox1t/7be1b459c71d87a10c21ffa73584097a
> * https://gist.github.com/fox1t/6206469e855c350ccf69c3e6255f4357
> 
> :)

Wow thanks a lot for this detailed explanation, so I was just looking it wrong all this time trying to combine both into one `type`. I appreciate the time spent for sharing this, going to update the typings

**Review by fox1t — APPROVED** [review]
LGTM

**Comment by fox1t** [comment]
> > @gkampitakis I figured out how to accomplish what we need.
> > Some points to remember:
> > 
> > 1. if a function returns `void` the implementer can return every type and the checks will pass. You can see it as `any` but you can't assign it. It exists for the caller and it literally means "You should don't care about the value my returned value"
> > 2. we need to have 2 different interfaces one for the async version and the other for the callback one. In fact, overloads are intersections and not unions and the passed function should implement both versions at the same time (this is impossible).
> >    I created 2 gists, feel free to use them to "adjust" the typings.
> > 
> > 
> > * https://gist.github.com/fox1t/7be1b459c71d87a10c21ffa73584097a
> > * https://gist.github.com/fox1t/6206469e855c350ccf69c3e6255f4357
> > 
> > :)
> 
> Wow thanks a lot for this detailed explanation, so I was just looking it wrong all this time trying to combine both into one `type`. I appreciate the time spent for sharing this, going to update the typings

No worries! TS can be really hard to understand sometimes. XD

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**